Mathlib/Topology/Category/Profinite/Light.lean

```diff
@@ -0,0 +1,313 @@
+  isLimit := {
+    lift := fun s ↦ s.π.app ⟨0⟩
+    fac := fun s j ↦ (s.π.naturality (homOfLE (zero_le (unop j))).op)
+    uniq := fun _ _ h ↦  h ⟨0⟩ }
+
+noncomputable
+def LightProfinite.of (F : ℕᵒᵖ ⥤ FintypeCat) : LightProfinite where
+  diagram := F
+  isLimit := limit.isLimit (F ⋙ FintypeCat.toProfinite)
+
+class Profinite.IsLight (S : Profinite) : Prop where
+  countable_clopens : Countable {s : Set S // IsClopen s}
+
+attribute [instance] Profinite.IsLight.countable_clopens
+
+open Classical in
+instance (S : Profinite) [S.IsLight] : Countable (DiscreteQuotient S) := by
+  have : ∀ d : DiscreteQuotient S, Fintype d := fun d ↦ Fintype.ofFinite _
+  refine @Function.Surjective.countable ({t : Finset {s : Set S // IsClopen s} //
+    (∀ (i j : {s : Set S // IsClopen s}), i ∈ t → j ∈ t → i ≠ j → i.val ∩ j.val = ∅) ∧
+    ∀ (x : S), ∃ i, i ∈ t ∧ x ∈ i.val}) _ _ ?_ ?_
+  · intro t
+    refine ⟨⟨fun x y ↦ ∃ i, i ∈ t.val ∧ x ∈ i.val ∧ y ∈ i.val, ⟨by simpa using t.prop.2,
+      fun ⟨i, h⟩ ↦ ⟨i, ⟨h.1, h.2.2, h.2.1⟩⟩, ?_⟩⟩, ?_⟩
+    · intro x y z ⟨ixy, hxy⟩ ⟨iyz, hyz⟩
+      refine ⟨ixy, hxy.1, hxy.2.1, ?_⟩
+      convert hyz.2.2
+      by_contra h
+      have hh := t.prop.1 ixy iyz hxy.1 hyz.1 h
+      apply Set.not_mem_empty y
+      rw [← hh]
+      exact ⟨hxy.2.2, hyz.2.1⟩
+    · intro x
+      simp only [setOf, Setoid.Rel]
+      obtain ⟨i, h⟩ := t.prop.2 x
+      convert i.prop.1 with z
+      refine ⟨fun ⟨j, hh⟩ ↦ ?_, fun hh ↦ ?_⟩
+      · suffices i = j by rw [this]; exact hh.2.2
+        by_contra hhh
+        have hhhh := t.prop.1 i j h.1 hh.1 hhh
+        apply Set.not_mem_empty x
+        rw [← hhhh]
+        exact ⟨h.2, hh.2.1⟩
+      · exact ⟨i, h.1, h.2, hh⟩
+  · intro d
+    refine ⟨⟨(Set.range (fun x ↦ ⟨d.proj ⁻¹' {x}, d.isClopen_preimage _⟩)).toFinset, ?_, ?_⟩, ?_⟩
+    · intro i j hi hj hij
+      simp only [Set.toFinset_range, Finset.mem_image, Finset.mem_univ, true_and] at hi hj
+      obtain ⟨ai, hi⟩ := hi
+      obtain ⟨aj, hj⟩ := hj
+      rw [← hi, ← hj]
+      dsimp
+      ext x
+      refine ⟨fun ⟨hhi, hhj⟩ ↦ ?_, fun h ↦ by simp at h⟩
+      simp only [Set.mem_preimage, Set.mem_singleton_iff] at hhi hhj
+      exfalso
+      apply hij
+      rw [← hi, ← hj, ← hhi, ← hhj]
+    · intro x
+      refine ⟨⟨d.proj ⁻¹' {d.proj x}, d.isClopen_preimage _⟩, ?_⟩
+      simp
+    · ext x y
+      simp only [DiscreteQuotient.proj, Set.toFinset_range, Finset.mem_image, Finset.mem_univ,
+        true_and, exists_exists_eq_and, Set.mem_preimage, Set.mem_singleton_iff, exists_eq_left',
+        Quotient.eq'']
+      exact ⟨d.iseqv.symm , d.iseqv.symm⟩
+
+instance (S : Profinite) : IsCofiltered (DiscreteQuotient S) := inferInstance
+
+noncomputable
+def cofinal_sequence (S : Profinite) [S.IsLight] : ℕ → DiscreteQuotient S := fun n ↦ by
+  induction n with
+  | zero => exact (exists_surjective_nat _).choose 0
+  | succ n D => exact
+      (IsCofiltered.toIsCofilteredOrEmpty.cone_objs ((exists_surjective_nat _).choose n) D).choose
```

Comment:
Does the equation compiler work here?

Reply:
You mean something like this? 

```
noncomputable
def cofinal_sequence (S : Profinite) [S.IsLight] : ℕ → DiscreteQuotient S := fun
  | .zero => (exists_surjective_nat _).choose 0
  | .succ n => (IsCofiltered.toIsCofilteredOrEmpty.cone_objs ((exists_surjective_nat _).choose n) 
      (cofinal_sequence S n)).choose
```

That works, I initially had some problems with getting it to work which is why it's in tactic mode now